src/runtime_src/core/pcie/tools/cloud-daemon/container/container.cpp

```diff
@@ -0,0 +1,235 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <errno.h>
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/file.h>
+#include <syslog.h>
+
+#include <cstdio>
+#include <cstring>
+#include <cassert>
+#include <algorithm>
+#include <stdlib.h>
+#include <thread>
+#include <chrono>
+#include <iostream>
+#include <fstream>
+#include <uuid/uuid.h>
+#include <openssl/md5.h>
+#include "xclbin.h"
+#include "container.h"
+
+/*
+ * Functions each plugin needs to provide
+ */
+extern "C" {
+int init(mpd_plugin_callbacks *cbs);
+void fini(void *mpc_cookie);
+}
+
+/*
+ * Init function of the plugin that is used to hook the required functions.
+ * The cookie is used by fini (see below). Can be NULL if not required.
+ */ 
+int init(mpd_plugin_callbacks *cbs)
+{
+	int ret = 1;
+	auto total = pcidev::get_dev_total();
+	if (total == 0) {
+		syslog(LOG_INFO, "Container: no device found");
+		return ret;
+	}
+    if (cbs) 
+	{
+		// hook functions
+		cbs->mpc_cookie = NULL;
+		cbs->get_remote_msd_fd = get_remote_msd_fd;
+		cbs->load_xclbin = xclLoadXclBin;
+		ret = 0;
+	}
+    syslog(LOG_INFO, "container mpd plugin init called: %d\n", ret);
+    return ret;
+}
+
+/*
+ * Fini function of the plugin
+ */ 
+void fini(void *mpc_cookie)
+{
+     syslog(LOG_INFO, "container mpd plugin fini called\n");
+}
+
+/*
+ * callback function that is used to setup communication channel
+ * we are going to handle mailbox ourself, no comm channel is required.
+ * so just return -1 to the fd
+ * Input:
+ *		d: dbdf of the user PF
+ * Output:    
+ *		fd: socket handle of the communication channel
+ * Return value:
+ *		0: success
+ *		1: failure
+ */ 
+int get_remote_msd_fd(size_t index, int& fd)
+{
+	fd = -1;
+	return 0;
+}
+
+/*
+ * callback function that is used to handle MAILBOX_REQ_LOAD_XCLBIN msg
+ * 
+ * Input:
+ *		index: index of the FPGA device
+ *		xclbin: the fake xclbin file
+ * Output:   
+ *		none	
+ * Return value:
+ *		0: success
+ *		1: failure
+ */ 
+int xclLoadXclBin(size_t index, const axlf *&xclbin)
+{
+	auto d = std::make_unique<Container>(index);
+	if (!d->isGood())
+		return 1;
+	return d->xclLoadXclBin(xclbin);
+}
+
+int Container::xclLoadXclBin(const xclBin *&buffer)
+{
+	/*
+	 * This file is delivered by default not to provide xclbin protection.
+	 * That means, the input xclbin file is downloaded as-is. This is also
+	 * useful in xrt container deployment case, with which, only the user PF
+	 * is granted to the container, and mgmt is left at host, but within
+	 * container users can use xbutil and/or openCL without seeing any
+	 * difference compared to within host.
+	 *
+	 * If container platform vendors, say Nimbix, want to have xclbin proction,
+	 * their code can be added here.
+     */	 
+#if 1
```

Comment:
remove?

Reply:
I made it on purpose. Better to keep. 
#if 1, is the default, return original xclbin
#else, is what my sample code below. And it is used for my test and for user's reference as well